rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,363 @@
+    
+    // The maximum timestamp of any row in the SST
+    max_ts: long;
+}
+```
+
+This value will be written when the SST is created. We will read this back using a codec that maps
+1:1 with the format version. The version 0 has the following per-value format:
+
+```
+|  u16    | var |    u32    |  var  |
+|---------|-----|-----------|-------|
+| key_len | key | value_len | value |
+|---------|-----|-----------|-------|
+```
+
+The `format_version` will be bumped to 1 and will be modified to have the following schema:
+
+```
+|  u16    | var | var  |    u32    |  var  |
+|---------|-----|------|-----------|-------|
+| key_len | key | meta | value_len | value |
+|---------|-----|------|-----------|-------|
+```
+
+The newly introduced `meta` field will be decoded using the `meta_features` array specified for
+this SST. Which values meta features are included depend on the data that is being written. If
+any row in the SST has a `ttl`, the `TimeToLive` feature will be enabled.
+
+As part of this proposal, we will no longer represent tombstones as `INT_MAX` value_len and instead
+have an indicator `byte` in the meta array which specifies the `RowFlags`. `RowFlags` is a bitmask
+of descriptors on the row. The first bit in `RowFlags` is set to `1` if the row is a tombstone.
+Other bits are reserved for future flags. Later we can add other types of metadata rows as
+necessary (since flatbuffers does not support single-bit values, a byte is the smallest amount of
+data we can store here):
+
+```
+                |-- meta --|
+|  u16    | var |   byte   |    u32    |  var  |
+|---------|-----|----------|-----------|-------|
+| key_len | key | row_type | value_len | value |
+|---------|-----|----------|-----------|-------|
+```
+
+In cases where both `Timestamp` and `TimeToLive` are enabled, the row will look like:
+
+```
+                |-------- meta --------------|
+|  u16    | var |    byte  | i64 |    i64    |    u32    |  var  |
+|---------|-----|----------------------------|-----------|-------|
+| key_len | key | row_type | ts  | expire_ts | value_len | value |
+|---------|-----|----------|-----|-----------|-----------|-------|
+```
+
+Both `ts` and `expire_ts` are encoded as `i64` in milliseconds.
+
+The metadata will be returned in the decoded `KeyValueDeletable` which will be modified:
+
+```rust
+pub struct KeyValueDeletable {
+    pub key: Bytes,
+    pub value: ValueDeletable,
+    pub meta: KeyValueMeta,
+}
+
+pub struct KeyValueMeta {
+    pub ts: Option<i64>,
+    pub expire_ts: Option<i64>,
+}
+```
+
+### Compaction
+
+#### TTL Compaction Filter
+
+We will introduce the concept of a compaction filter, which serves as a mechanism for pluggable
+participation in the compaction process. Specifically, the filter (if configured) will run on each
+key during the compaction process and optionally remove or modify the value.
+
+For the scope of this RFC, we will be introducing a builtin TS/TTL compaction filter. This filter
```

Comment:
I think this compaction filter will run for every compaction, and is independent of `periodic_compaction_interval` added in next section. Can this be stated explicitly?